bin/commands/start.py

```diff
@@ -0,0 +1,94 @@
+import os
+import re
+
+import utils
+
+DESCRIPTION = "Start a container from an existing dhis2-data Docker image or from an exported file"
+
+
+def setup(parser):
+    parser.add_argument(
+        "image_or_file", metavar="IMAGE_OR_EXPORT_FILE", help="Docker image or exported file"
+    )
+    utils.add_core_image_arg(parser)
+    parser.add_argument(
+        "-d", "--detach", action="store_true", help="Run containers on the background"
+    )
+    parser.add_argument(
+        "-k", "--keep-containers", action="store_true", help="Keep existing containers"
+    )
+    parser.add_argument("--run-sql", metavar="DIRECTORY", help="Run .sql[.gz] files in directory")
+    parser.add_argument(
+        "--run-scripts",
+        metavar="DIRECTORY",
+        help="Run shell scripts in directory (default: PRE tomcat; if filename starts with 'post': POST tomcat)",
+    )
+    parser.add_argument("--pull", action="store_true", help="Force a pull from docker hub")
+    parser.add_argument("-p", "--port", type=int, metavar="N", help="Set Dhis2 instance port")
+
+
+def run(args):
+    image_or_file = args.image_or_file
+
+    if os.path.exists(image_or_file) and os.path.isfile(image_or_file):
+        image_name = import_from_file(image_or_file)
+        start(args, image_name)
+    else:
+        start(args, image_or_file)
+
+
+def import_from_file(images_path):
+    dhis2_data_image_re = "/{}:".format(utils.DHIS2_DATA_IMAGE)
+    result = utils.load_images_file(images_path)
+    lines = result.stdout.decode("utf-8").splitlines()
+    lines_splitted = [line.split() for line in lines]
+    dhis2_data_images = [
+        parts[-1] for parts in lines_splitted if parts and re.search(dhis2_data_image_re, parts[-1])
+    ]
+    if dhis2_data_images:
+        return dhis2_data_images[0]
+    else:
+        msg = "Cannot find dhis2 data image (pattern={})".format(dhis2_data_image_re)
+        raise utils.D2DockerError(msg)
+
+
+def start(args, image_name):
+    utils.logger.info("Start image: {}".format(image_name))
+    result = utils.get_image_status(image_name)
+    if result["state"] == "running":
+        msg = "Container already runnning for image {}".format(result["containers"]["db"])
+        raise utils.D2DockerError(msg)
+    port = args.port or utils.get_free_port()
+    utils.logger.info("Port: {}".format(port))
+    core_image = args.core_image
+
+    if args.pull:
+        utils.run_docker_compose(["pull"], image_name, core_image=core_image)
+    if args.keep_containers and args.run_sql:
+        raise utils.D2DockerError("--run-sql cannot be used with --keep-containers")
```

Comment:
Not sure about this exception. I normally seed my database on my current dockerized environment and I reseed it to apply changes way before the initial seeding. I have smaller SQL files to change user passwords such as the admin or clean the analytics tables.

I would instead show a warning but allow the user to proceed. If not comfortable with that happening as default at least some way to force it with an additional arg.

Reply:
This restriction is not by design, but forced by how the init script of the postgres image works. The initialization (which also sets up postgis) is done only the very first time, when the PGDATA is not created:

https://github.com/docker-library/postgres/blob/master/10/docker-entrypoint.sh#L57

I was thinking: maybe we can remove the SQL seeding from here and move it to the core script (we will need to install the psql client in that container, though). Also, this way the sql/shell seeding would be done in the same place and we'd have no restrictions.